src/core_plugins/kibana/public/dashboard/panel/dashboard_panel.js

```diff
@@ -0,0 +1,121 @@
+import React from 'react';
+import PropTypes from 'prop-types';
+import classNames from 'classnames';
+
+import { DashboardViewMode } from '../dashboard_view_mode';
+import { PanelHeader } from './panel_header';
+
+export class DashboardPanel extends React.Component {
+
+  constructor(props) {
+    super(props);
+    this.state = {};
+    this.embeddable = null;
+    this.embeddableHandler = null;
+    this._isMounted = false;
+  }
+
+  async componentDidMount() {
+    this._isMounted = true;
+    const { getEmbeddableHandler, panel, getContainerApi } = this.props;
+
+    this.containerApi = getContainerApi();
+    this.embeddableHandler = getEmbeddableHandler(panel.type);
+    const editUrl = await this.embeddableHandler.getEditPath(panel.id);
+    const title = await this.embeddableHandler.getTitleFor(panel.id);
+
+    // TODO: use redux instead of the isMounted anti-pattern to handle the case when the component is unmounted
+    // before the async calls above return. We can then get rid of the eslint disable line. Without redux, there is
+    // not a better option, since you aren't supposed to run async calls inside of componentWillMount.
+    if (this._isMounted) {
+      /* eslint-disable react/no-did-mount-set-state */
+      this.setState({ editUrl, title });
+
+      this.destroyEmbeddable = await this.embeddableHandler.render(
```

Comment:
Why are we waiting for the title/editUrl to be available before calling `this.embeddableHandler.render` they don't seem to be dependent on each other? If we did the above suggestion, and queued off this async request immediately, I'm not seeing the reason for us to track the `_isMounted`, but we could just use the `this.destroyEmbeddable !== null)` check directly in the `componentDidUnmount` call.

Reply:
I need isMounted check before the setStates or I can get:
`Warning: setState(...): Can only update a mounted or mounting component. This usually means you called setState() on an unmounted component. This is a no-op. Please check the code for the DashboardPanel component.`

I can still use your above suggestion but I don't think I can get rid of isMounted.  Instead it will look like this:

```
    this.embeddableHandler.getEditPath(panel.id).then(editUrl => {
      if (this._isMounted) { this.setState({ editUrl }); }
    });

    /* eslint-disable react/no-did-mount-set-state */
    this.embeddableHandler.getTitleFor(panel.id).then(title => {
      if (this._isMounted) { this.setState({ title }); }
    });

    this.destroyEmbeddable = await this.embeddableHandler.render(
      this.panelElement,
      panel,
      this.containerApi);
```

fwiw, the redux variation gets rid of all this.  I actually changed this.embeddableHandler.render to return an Embeddable, which contains state for the rendered embeddable, including title and editUrl - so there is only one async call, not three - https://github.com/elastic/kibana/pull/14103/files#diff-a18d16d17e2c646a166a4d5523608fbfR86/

> Why are we waiting for the title/editUrl to be available before calling this.embeddableHandler.render

Wasn't intentional, I just liked the await syntax.  Updated to your suggestion.

Reply:
Ah, gotcha. Pretty excited about the redux changes coming, that makes sense.